api/renter.go

```diff
@@ -350,49 +387,156 @@ func (api *API) renterDeleteHandler(w http.ResponseWriter, req *http.Request, ps
 
 // renterDownloadHandler handles the API call to download a file.
 func (api *API) renterDownloadHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	destination := req.FormValue("destination")
-	// Check that the destination path is absolute.
-	if !filepath.IsAbs(destination) {
-		WriteError(w, Error{"destination must be an absolute path"}, http.StatusBadRequest)
+	p, errmsg := api.parseAndValidateDownloadParameters(w, req, ps)
+	if errmsg != nil {
+		WriteError(w, *errmsg, http.StatusBadRequest)
 		return
 	}
 
-	err := api.renter.Download(strings.TrimPrefix(ps.ByName("siapath"), "/"), destination)
-	if err != nil {
-		WriteError(w, Error{"download failed: " + err.Error()}, http.StatusInternalServerError)
-		return
+	if p.Async { // Create goroutine if `async` param set.
+		go api.renter.DownloadSection(p)
+	} else {
+		err := api.renter.DownloadSection(p)
+		if err != nil {
+			WriteError(w, Error{"download failed: " + err.Error()}, http.StatusInternalServerError)
+			return
+		}
 	}
 
-	WriteSuccess(w)
+	if !p.Httpresp {
+		WriteSuccess(w)
+	}
 }
 
 // renterDownloadAsyncHandler handles the API call to download a file asynchronously.
 func (api *API) renterDownloadAsyncHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	destination := req.FormValue("destination")
-	// Check that the destination path is absolute.
-	if !filepath.IsAbs(destination) {
-		WriteError(w, Error{"destination must be an absolute path"}, http.StatusBadRequest)
+	// Ensure that the async flag is not set to false.
+	if strings.Contains(req.URL.RawQuery, "async=false") {
+		WriteError(w, Error{"/downloadasync does not support allow requests containing `async=false`."}, http.StatusBadRequest)
 		return
 	}
 
-	siapath := strings.TrimPrefix(ps.ByName("siapath"), "/")
-	fileExists := false
+	// Set async flag to true.
+	req.URL.RawQuery += "&async=true"
+	p, errmsg := api.parseAndValidateDownloadParameters(w, req, ps)
+	if errmsg != nil {
+		WriteError(w, *errmsg, http.StatusBadRequest)
+		return
+	}
+
+	go api.renter.DownloadSection(p)
+
+	WriteSuccess(w)
+}
```

Comment:
you gave an example of a `downloadinorder` flag. We'd handle that by defaulting the flag to `true` (meaning it's only false if explicitly written false), and then making sure that if the user sets the flag to false, they don't set it along side http streaming. Out-of-order downloads just don't make sense if you are doing streaming (it's possible, but would overly complicate the api).